src/app/GroupLayer.tsx

```diff
@@ -0,0 +1,272 @@
+/** @jsxImportSource @emotion/react */
+
+// This file contains code related to the group step. Earlier, we had
+// code related to each step spread across multiple files, the reason
+// was that we architected the intial technical spike around tools.
+// Now, we have a single file for each step. Eventhough we have a lot
+// of repetitive code because of this change, eventually, we will
+// refactor and fix it.
+
+import { Dispatch } from "redux";
+import { GroupLayerActionMenu } from "../components/ActionMenu";
+import Annotation, {
+  AnnotationBeingCreated,
+  AnnotationProps,
+  CreationState,
+  TranslucentBox,
+} from "./Annotation";
+import { CreateAnnotationAttr, NO_OP, RenderAnnotationsHandler } from "./PDF";
+import {
+  useSelector,
+  useDispatch,
+  AccessibleForm,
+  AnnotationId,
+} from "./StoreProvider";
+import React from "react";
+import Xarrow from "react-xarrows";
+import { AllTokens } from "./LabelLayer";
+
+export const groupLayerHandlers = (
+  state: AccessibleForm,
+  dispatch: Dispatch,
+  createAnnotationAttr: CreateAnnotationAttr
+) => {
+  const {
+    div: container,
+    creationState,
+    newCreationBounds,
+    resetCreationState,
+    updateCreationState,
+  } = createAnnotationAttr;
+  const { tool, selectedAnnotations } = state;
+
+  switch (tool) {
+    case "CREATE": {
+      return {
+        cursor: "crosshair",
+        creationState,
+        container,
+        onClick: NO_OP,
+        onMouseDown: newCreationBounds,
+        onMouseMove: updateCreationState,
+        onMouseLeave: resetCreationState,
+        onMouseUp: () => {
+          if (!creationState) return;
+          // The following line stops dispatch of CREATE_ANNOTATION_FROM_TOKENS if no token is selected.
+          // It fixes the bug where, an action is dispatched without tokens, empty token created a new
+          // annotation whose dimeasions are defined by boxContaning function(as Math.MAX) in utils.
+          if (creationState.tokens.length === 0) return;
+          const id = window.crypto.randomUUID();
+          dispatch({
+            type: "CREATE_ANNOTATION_FROM_TOKENS",
+            payload: {
+              ui: {
+                id,
+                backgroundColor: "rgb(36, 148, 178, 0.4)",
+                border: "3px solid rgb(36, 148, 178)",
+                type: "GROUP_LABEL",
+              },
+              tokens: creationState.tokens,
+            },
+          });
+          dispatch({
+            type: "CREATE_LABEL_RELATION",
+            payload: { from: Object.keys(selectedAnnotations)[0], to: id },
+          });
+          resetCreationState();
+          // As soon as a label is created, we switch user to the select tool.
+          dispatch({
+            type: "CHANGE_TOOL",
+            payload: "SELECT",
+          });
+          // After label is created we no longer want annotation to be selected.
```

Comment:
In general, I don't think an `onClick` should have more than one call to `dispatch` at a time. Rather than dispatching several actions, if we dispatch one action and handle all of the state changes in the reducer, that can simplify a lot of the logic.